### ISSUE TITLE
Add special fields `updatedAt` and `createdAt` of type DateTime

### ISSUE THREAD

**Issue by jesstelford** [keystonejs/keystone-5#184]
Very useful fields when doing sorting (eg, for a news feed)

## `createdAt`

Is set when a new entry is made to the current datetime. It cannot be updated, only read.

## `updatedAt`

Is set every time any values in the item are changed. It cannot be directly updated, only indirectly via an `update` mutation. It can be read.

Initial value == `createdAt`

**Comment by jesstelford** [comment]
Two possible APIs:

```javascript
import { withTrackingFields } from '@keystone-alpha/tracking-fields';
keystone.createList('Post', withTrackingFields({
  fields: {
    title: { type: Text },
  },
}));
```

```javascript
import { buildTrackingFields } from '@keystone-alpha/tracking-fields';
keystone.createList('Post', {
  fields: {
    title: { type: Text },
    ...buildTrackingFields(),
  },
});
```

I like the first (`withTrackingFields`) a bit better as it can do more things, like injecting `hooks` / `access` at the list level rather than only at the field level.

**Comment by gautamsi** [comment]
> For prior art, see next's way of modifying their config object.

I am not against higher order functions but there is more complexity here than React or Next.js. 
1. it will look ugly if we have to use multiple higher order functions with multiple arguments for them. 
2. It will be limited to narrow use case. with narrow I mean, it will be focused with specific tasks but not able to create a bigger scenario of extension/add-ins

I propose api like this:

```js
import { withTrackingFields, withPublishingFields } from '@keystone-alpha/tracking-fields';
keystone.createList('Post', {
  fields: {
    title: { type: Text },
  },
},
modifiers: [withTrackingFields, withPublishingFields], 
// or
extensions: [withTrackingFields, withPublishingFields],
); 
```

or using extension with keystone instance, I should use,

```js
import { PublishingExtension, TrackingExtension } from '@keystone-alpha/extensions';
const keystone = new Keystone({
  name: 'Keystone Demo Blog',
  adapter: new MongooseAdapter(),
  extensions: [PublishingExtension, TrackingExtension], // I would prefer instance here so that custom parameters can be passed
});
```

## Change needed

Keystone should run list options (1st and 2nd arguments) with modifiers/extensions internally. This way it is clean reusable with multiple modifiers.

in Extension scenario, we can have conventions to have specific methods (static or instance) on the Extension so that keystone can run each of them with required parameters at specific times like before calling `initFields` we run field option with `Extension.initFields`.

## Advantages of:

### using `modifiers/extensions: [withTrackingFields, withPublishingFields]`
* can use multiple modifiers, they can come from userland as well.
* modifiers can do much more than just adding its fields, like setup some internal list/field hooks. 

### using ` extensions: [PublishingExtension, TrackingExtension],`

* userland can bring more features which can be used for things impossible without forking otherwise. 
* userland can also change core behavior (can not think of anything specific atm but userland can be creative)


## Extension Points:
```js
import { PublishingExtension, TrackingExtension } from '@keystone-alpha/extensions';
// TrackingExtension has instanceMethod `initFieldExtension` or more appropriate name
// PublishingExtension has static methods `initFieldExtension` 
const trackingExtension = new TrackingExtension({
  timeTracking: true, // adds createdAt and updatedAt
  userTracking: true, // adds createdBy and updatedBy
})

const keystone = new Keystone({
  name: 'Keystone Demo Blog',
  adapter: new MongooseAdapter(),
  extensions: [PublishingExtension, trackingExtension], 
});
```
before following line we should run `trackingExtension.initFieldExtension(list)` or more suitable place before `initFields` run on list instance.

https://github.com/keystonejs/keystone-5/blob/1b8cb17979f8b458aebee567c3ef7edc9c803b71/packages/keystone/Keystone/index.js#L85

another advantages of using extension points that it can have collective hooks for specific purpose so that no individual passing of hooks for each list, each hook action.
 
similarly there can be other extension points for other things.

**Comment by jesstelford** [comment]
I like the idea of putting some config (whatever shape it takes) on the `keystone` instance level, and therefore having it apply broadly to all lists 👍 

The specific problem in this issue I think is better served _without_ adding the complexity of a plugin/extension system, but I would like to see what an extension system could shake out to be in the future (and we could eventually fold a `createdAt`/`createdBy` into that extension system if we go with another API for now).

**Comment by gautamsi** [comment]
its not that complex, Let me create a POC branch for you to review. 

If we add something now in `alpha` it will have better chance to evolve before GA. adding extension point early will help more otherwise we will have to wait for next major version. 

Another use case is that we add extension support now, it will be easy for others to build upon without having to wait for next release on npm. It will also evolve with internal changes to add more extension point without breaking existing app which uses no extensions.

**Comment by gautamsi** [comment]
@jesstelford 

This is what I came up with, I created both, `withTracking` which is easy and also my proposed modifier way, with little change to `keystone.createList`. This change is rudimentary just for POC, there can be much bigger scenario for this. 

I have this setup in #998 for review. thoughts?

## Setup
### Additional fields by

`withTracking` higher order function.

`timeTracking` (I am not good at naming them yet!) modifier, this version of the modifier also fires if there is any existing `beforeChange` hook on the list.

`timeTracking2` modifier, this version of the modifier replaces the hooks for `beforeChange` and do not fires if there is any existing `beforeChange` hook on the list.

#### Lists changed
`Post` list is setup with `withTracking` as well as `timeTracking` 

`PostCategory` list is setup with `withTracking` as well as `timeTracking2`

Difference in higher order functions and the modifiers are used. in fact the higher order function can be a modifier with little change (remove unused `listKey` in calling modifier)

```js
// blog demo
// Higher order functions
keystone.createList('Post', withTracking(Post))

// modifier
keystone.createList('Post', Post, { modifiers: timeTracker });
// with multiple modifiers
keystone.createList('Post', withTracking(Post), { modifiers: [timeTracker, timeTracker2] });
```

I can say it's not that ugly after @JedWatson refactored list configs in `schema.js` for blog demo. It will still look wired to have multiple higher order functions used inline. much cleaner when you use modifiers option in 3rd argument.



unfortunately your second suggestion of `...buildTrackingFields(),` may not work as it can not setup list hooks necessary for this scenario.


### Result in GraphiQL

Example `create{List}` mutation

```gql
mutation {
  createPost(data: { title: "test", body: "test" }) {
    id
    title
    status
    createdAt
    updatedAt
    whenCreated
    whenUpdated
  }

  createPostCategory(data: { name: "test", slug: "test" }) {
    name
    slug
    createdAt
    whenCreated
    whenUpdated
  }
}
```
This is what I get in return, notice that `CreatedAt` (and `updatedAt` in update mutation) for `PostCategory` did not get updated as modifier did not run the original mutation. hereby I propose making hooks an array which should be appended.
```gql
{
  "data": {
    "createPost": {
      "id": "5ca93731f77d99f87f009c85",
      "title": "test",
      "status": "draft",
      "createdAt": "2019-04-06T23:33:05.315Z",
      "updatedAt": null,
      "whenCreated": "2019-04-06T23:33:05.315Z",
      "whenUpdated": null
    },
    "createPostCategory": {
      "id": "5ca93731f77d99f87f009c86",
      "name": "test",
      "slug": "test",
      "createdAt": null,
      "updatedAt": null,
      "whenCreated": "2019-04-06T23:33:05.321Z",
      "whenUpdated": null
    }
  },
  "extensions": {}
}
```

Example of `update{List}` mutation
```gql
mutation {
  updatePost(id: "5ca936c9f77d99f87f009c83"
    data: { title: "test3" }) {
    id
    title
    createdAt
    updatedAt
    whenCreated
    whenUpdated
  }
  updatePostCategory(
    id: "5ca936c9f77d99f87f009c84"
    data: { name: "test3", slug: "test3" }
  ) {
    id
    name
    slug
    createdAt
    updatedAt
    whenCreated
    whenUpdated
  }
}
```

result:
```gql
{
  "data": {
    "updatePost": {
      "id": "5ca936c9f77d99f87f009c83",
      "title": "test3",
      "createdAt": "2019-04-06T23:31:21.084Z",
      "updatedAt": "2019-04-06T23:36:01.239Z",
      "whenCreated": "2019-04-06T23:31:21.084Z",
      "whenUpdated": "2019-04-06T23:36:01.239Z"
    },
    "updatePostCategory": {
      "id": "5ca936c9f77d99f87f009c84",
      "name": "test3",
      "slug": "test3",
      "createdAt": null,
      "updatedAt": null,
      "whenCreated": "2019-04-06T23:31:21.096Z",
      "whenUpdated": "2019-04-06T23:36:01.248Z"
    }
  },
  "extensions": {}
}
```

thoughts.

**Comment by gautamsi** [comment]
@jesstelford Do you have any comment for the changes I suggested? It would be helpful. if not, let me know if you are ok with implementation at https://github.com/keystonejs/keystone-5/pull/998/files#diff-a0df9e1c74610822f5bb840d5bd0a27b I can complete this and get to `createdBy` and `updatedBy`.

**Comment by gautamsi** [comment]
@jesstelford any comment ^^

**Comment by molomby** [comment]
I believe this can now be implemented using list plugins #1491.

**Comment by gautamsi** [comment]
@molomby #1291 is now using #1491 approach. still some blockers, see https://github.com/keystonejs/keystone-5/pull/1291#discussion_r313995849 and https://github.com/keystonejs/keystone-5/issues/1495
